library/oneview_server_profile.py

```diff
@@ -16,236 +16,293 @@
-            # ignore and try again
-            # This waiting time was chosen empirically and it could differ according to the hardware.
-            time.sleep(10)
-            pass
-
-    raise Exception("Could not allocate server hardware")
-
-
-def delete_profile(con, server_profile):
-    servers = hpov.servers(con)
-    servers.remove_server_profile(server_profile)
-
-
-def gather_facts(con, server_profile):
-    facts = {}
-    facts['serial_number'] = server_profile['serialNumber']
-    facts['server_profile'] = server_profile
-    facts['server_hardware'] = con.get(server_profile['serverHardwareUri'])
-    srv = hpov.servers(con)
-    facts['compliance_preview'] = srv.get_server_profile_compliance(server_profile)
-    facts['created'] = False
-    return facts
-
-
-def make_compliant(con, server_profile):
-    changed = False
-    srv = hpov.servers(con)
-
-    changed = False
-    if (server_profile['templateCompliance'] != 'Compliant'):
-        # check if server can be remediated while powered on
-        srv.update_server_profile_from_template(server_profile)
-        changed = True
-
-    return changed
-
-
-def set_power_state(con, server_profile, power_state):
-    srv = hpov.servers(con)
-    server_hardware_uri = server_profile['serverHardwareUri']
-    server_hardware = con.get(server_hardware_uri)
-    power_state_mapping = {'powered_on': 'On', 'powered_off': 'Off'}
-    srv.set_server_powerstate(server_hardware, power_state_mapping[power_state], True)
+            else:
+                if state in ["powered_on", "powered_off"]:
+                    self.module.fail_json(msg="Cannot find server to put in state : " + state)
+                # we didnt find an existing one, so we create a profile
+                elif state in ['present']:
+                    server_profile = self.create_profile(server_name, server_template)
+                    facts = self.gather_facts(server_profile)
+                    facts['created'] = True
+                    self.module.exit_json(
+                        changed=True, msg='Created profile', ansible_facts=facts
+                    )
+        except Exception as e:
+            self.module.fail_json(msg=e.message)
+
+    def update_profile(self, server_profile, server_template):
+        """ update the server to match the template """
+        changed = False
+        if (server_profile['serverProfileTemplateUri'] != server_template['uri']):
+            server_profile['serverProfileTemplateUri'] = server_template['uri']
+            self.oneview_client.server_profiles.update(server_profile)
+            changed = True
+
+        return changed
+
+    def create_profile(self, server_name, server_template):
+
+        # find servers that have no profile, powered off mathing Server hardware type
+        server_hardware_name = self.module.params['server_hardware']
+        tries = 0
+        while tries < 25:
```

Comment:
This loop exists because the hardware is automatically picked up if not informed, so if I run a script to create a profile server in more than one host, the creation will be concurrent and both profiles will try to get the same hardware. This number has been chosen empirically  to address this problem